pkg/services/datasources/datasource_initializer.go

```diff
@@ -0,0 +1,101 @@
+package datasources
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana/pkg/bus"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/plugins/adapters"
+	"github.com/grafana/grafana/pkg/plugins/manager"
+)
+
+const (
+	adminUserID = 1
+	orgID       = 1
+)
+
+var logger = log.New("datasource_initializer")
+
+func ProvideInitializerService(dataSourceCache CacheService, plugReqValidator models.PluginRequestValidator,
+	pm plugins.Manager,
+	pluginManager *manager.PluginManager) *DataSourceInitializerService {
+	return &DataSourceInitializerService{
+		DataSourceCache:    dataSourceCache,
+		PluginManager:      pm,
+		OtherPluginManager: pluginManager,
+	}
+}
+
+type DataSourceInitializerService struct {
+	DataSourceCache    CacheService
+	PluginManager      plugins.Manager
+	OtherPluginManager *manager.PluginManager
+}
+
+func (di *DataSourceInitializerService) Run(ctx context.Context) error {
+	query := models.GetDataSourcesQuery{OrgId: orgID, DataSourceLimit: -1}
+
+	if err := bus.Dispatch(&query); err != nil {
+		return fmt.Errorf("failed to query datasources: %v", err)
+	}
+
+	for _, ds := range query.Result {
+		err := di.CheckDatasourceHealth(ds.Id, ds.Name)
```

Comment:
@wbrowne @marefr -- any suggestions for better places to make an initial healthcheck call?

perhaps part of `restartKilledProcess`?